README.md
[![Build Status](https://travis-ci.org/kianmeng/webservice-pokemon.svg?branch=master)](https://travis-ci.org/kianmeng/webservice-pokemon)
[![Cpan license](https://img.shields.io/cpan/l/WebService-Pokemon.svg)](https://metacpan.org/release/WebService-Pokemon)
[![Cpan version](https://img.shields.io/cpan/v/WebService-Pokemon.svg)](https://metacpan.org/release/WebService-Pokemon)

# NAME

WebService::Pokemon - Perl library for accessing the Pokémon data,
http://pokeapi.co.

# SYNOPSIS

    use WebService::Pokemon;

    my $pokemon_api = WebService::Pokemon->new;

    # By id.
    my $pokemon = $pokemon_api->resource('berry', 1);

    # By name.
    my $pokemon = $pokemon_api->resource('berry', 'cheri');

# DESCRIPTION

WebService::Pokemon is a Perl client helper library for the Pokemon API (pokeapi.co).

# DEVELOPMENT

Source repository at [https://github.com/kianmeng/webservice-pokemon](https://github.com/kianmeng/webservice-pokemon).

How to contribute? Follow through the [CONTRIBUTING.md](https://github.com/kianmeng/webservice-pokemon/blob/master/CONTRIBUTING.md) document to setup your development environment.

# METHODS

## new(\[%$args\])

Construct a new WebService::Pokemon instance. Optionally takes a hash or hash reference.

    # Instantiate the class.
    my $pokemon_api = WebService::Pokemon->new;

### api\_url

The URL of the API resource.

    # Instantiate the class by setting the URL of the API endpoints.
    my $pokemon_api = WebService::Pokemon->new({api_url => 'http://example.com/api/v2'});

    # Or after the object was created.
    my $pokemon_api = WebService::Pokemon->new;
    $pokemon_api->api_url('http://example.com/api/v2');

### autoload

Set this if you want to expand all fields point to external URL.

    # Instantiate the class by setting the URL of the API endpoints.
    my $pokemon_api = WebService::Pokemon->new({autoload => 1});

    # Or after the object was created.
    my $pokemon_api = WebService::Pokemon->new;
    $pokemon_api->autoload(1);
    $pokemon_api->resource('berry');

## resource($resource, \[$name\], \[$limit\], \[$offset\])

Get the details of a particular resource with optional id or name; limit per
page, or offset by the record list.

    # Get paginated list of available berry resource.
    my $berry = $pokemon_api->resource('berry');

    # Or by page through limit and pagination.
    my $berry = $pokemon_api->resource('berry', undef, 60, 20);

    # Get by id.
    my $berry_firmness = $pokemon_api->resource('berry-firmnesses', 1);

    # Get by name.
    my $berry_firmness = $pokemon_api->resource('berry-firmnesses', 'very-soft');

## resource\_by\_url($url)

Get the details of a particular resource by full URL.

    # Get paginated list of available berry resource with default item size.
    my $berries = $pokemon_api->resource_by_url('https://pokeapi.co/api/v2/berry/');

    # Get paginated list of available berry resource with explicit default item size.
    my $berries = $pokemon_api->resource_by_url('https://pokeapi.co/api/v2/berry/?limit=20&offset=40');

    # Get particular berry resource.
    my $berry = $pokemon_api->resource_by_url('https://pokeapi.co/api/v2/berry/1');

# AUTHOR

Kian Meng, Ang <kianmeng@cpan.org>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2018-2019 by Kian Meng, Ang.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)


Changes
Revision history for WebService-Pokemon

{{$NEXT}}

0.11  2019-05-02 21:40:53 +08
        - Fix missing module version
        - Replace incorrect environment variable

0.10  2019-05-01 19:57:09 +08
        - Autoload results by choice
        - Remove CHI caching
        - Tidy up code using perltidy
        - Fix test due to API server changes
        - Fix perlcritic issues
        - Fix duplicate contributor names
        - Make live API server optionally

0.09  2018-06-10 00:45:49 +08
        - Abstract the web service responses in different data classes

0.08  2018-06-06 01:57:13 +08
        - Cache HTTP decoded content using CHI
        - Speed testing using local cached results

0.07  2018-06-06 00:20:59 +08
        - Switch Web::API to Role::REST::Client for easy customization

0.06  2018-05-16 22:56:23 +08
        - Remove unused Makefile.PL
        - Build and test using Docker

0.05  2018-05-16 21:09:02 +08
        - Fix missing strictures
        - Update synopsis

0.04  2018-05-14 21:54:16 +08
        - Add berry, berry_firmness, berry_flavor endpoints
        - Enable switching between v1 and v2 API version

0.03  2018-05-06 00:52:52 +08
        - Fix module name mismatch

0.02  2018-05-06 00:42:34 +08
        - Switch module name according to standard naming convention
        - Integration with Travis CI and other external badges

0.01  2018-05-05 13:21:19 +08
        - Initial release


t/05_api_resource_list.pm
use strict;
use warnings;
use utf8;

use Test::More;

use WebService::Pokemon;

my ($got, $expect, $url) = ('', '', '');

my $api = WebService::Pokemon->new;

$got = $api->resource('berry');
is(scalar %{$got->results->[0]}, 2, 'expect no autoload of each results items');

$api->autoload(1);
$got = $api->resource('berry');
is(scalar %{$got->results->[0]}, 14, 'expect autoload of each results item');

done_testing;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A module to access the Pok\x{e9}mon data through RESTful API from http://pokeapi.co",
  "AUTHOR" => "Kian Meng, Ang <kianmeng\@users.noreply.github.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "Webservice-Pokemon",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Webservice::Pokemon",
  "PREREQ_PM" => {
    "strictures" => 0,
    "utf8" => 0,
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Scalar::Util" => 0,
    "Test::More" => "0.94"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "File::Find" => 0,
  "File::Temp" => 0,
  "Module::Build" => "0.28",
  "Scalar::Util" => 0,
  "Test::More" => "0.94",
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/WebService/Pokemon/APIResourceList.pm
package WebService::Pokemon::APIResourceList;

use utf8;
use strictures 2;
use namespace::clean;

use Moo;
use Types::Standard qw(Any ArrayRef HashRef InstanceOf Int Str);
use Test::More;

our $VERSION = '0.09';

has api => (
    isa => InstanceOf['WebService::Pokemon'],
    is => 'rw',
);

has response => (
    isa => HashRef,
    is  => 'rw',
);

has count => (
    isa => Int,
    is  => 'rw',
    lazy => 1,
    default => sub {
        my $self = shift;
        return $self->response->{count};
    },
);

has previous => (
    isa => Any,
    is  => 'rw',
    lazy => 1,
    default => sub {
        my $self = shift;
        return $self->response->{previous};
    },
);

has next => (
    isa => Any,
    is  => 'rw',
    lazy => 1,
    default => sub {
        my $self = shift;
        return $self->response->{next};
    },
);

has results => (
    isa => ArrayRef,
    is  => 'rw',
    lazy => 1,
    builder => 1,
);

sub _build_results {
    my ($self) = @_;

    return $self->response->{results} if (!$self->api->autoload);

    my $urls = [map { $_->{url} } @{$self->response->{results}}];

    my $results = [];
    foreach my $url (@{$urls}) {
        push @{$results}, $self->api->resource_by_url($url);
    }

    return $results;
}

1;


lib/WebService/Pokemon/NamedAPIResource.pm
package WebService::Pokemon::NamedAPIResource;

use utf8;
use strictures 2;
use namespace::clean;

use Moo;
use Types::Standard qw(HashRef InstanceOf);

our $VERSION = '0.09';

has api => (
    isa => InstanceOf['WebService::Pokemon'],
    is => 'rw',
);

has response => (
    isa => HashRef,
    is  => 'rw',
);

sub BUILD {
    my ($self, $args) = @_;

    foreach my $arg (keys %{$args}) {
        $self->$arg($args->{$arg}) if (defined $args->{$arg});
    }

    foreach my $arg (keys %{$self->response}) {
        $self->meta->add_attribute(
            $arg => (is => 'rw', lazy => 1, builder => 1)
        );

        my $value = $self->response->{$arg};

        if (ref $value eq 'HASH') {
            $value = WebService::Pokemon::NamedAPIResource->new(
                api => $self->api,
                response => $value
            );
        }

        if (ref $value eq 'ARRAY') {
            my $list;
            foreach my $v (@{$value}) {
                push @{$list},
                    WebService::Pokemon::NamedAPIResource->new(
                        api => $self->api,
                        response => $v
                    );
            }
            $value = $list;
        }

        $self->$arg($value);
    }

    return $self;
}

1;


lib/WebService/Pokemon.pm
package WebService::Pokemon;

use 5.008_005;
use strictures 2;
use namespace::clean;
use utf8;

use Moo;
use Types::Standard qw(Bool Str);
use URI::Fast qw(uri);

use WebService::Pokemon::APIResourceList;
use WebService::Pokemon::NamedAPIResource;

with 'Role::REST::Client';

use constant DEFAULT_ITEMS_PER_PAGE => 20;
use constant DEFAULT_ITEMS_OFFSET => 0;

our $VERSION = '0.09';

has 'api_url' => (
    isa => Str,
    is => 'rw',
    default => sub { 'https://pokeapi.co/api/v2' },
);

has autoload => (
    isa => Bool,
    is  => 'rw',
    default => sub { 0 },
);

sub BUILD {
    my ($self, $args) = @_;

    foreach my $arg (keys %{$args}) {
        $self->$arg($args->{$arg}) if (defined $args->{$arg});
    }

    $self->set_persistent_header('User-Agent' => __PACKAGE__ . q| |
            . ($WebService::Pokemon::VERSION || q||));
    $self->server($self->api_url);

    return $self;
}

sub _request {
    my ($self, $resource, $id_or_name, $queries) = @_;

    return if (!defined $resource || length $resource <= 0);

    $queries ||= {};

    # In case the api_url was updated.
    $self->server($self->api_url);
    $self->type(q|application/json|);

    my $endpoint = q||;
    $endpoint .= qq|/$resource|;
    $endpoint .= qq|/$id_or_name| if (defined $id_or_name);

    my $response = $self->get($endpoint, $queries);
    my $response_data = $response->data;

    return $response_data;
}

sub resource {
    my ($self, $resource, $id_or_name, $limit, $offset) = @_;

    my $queries;
    if (!defined $id_or_name) {
        $queries->{limit} = $limit || DEFAULT_ITEMS_PER_PAGE;
        $queries->{offset} = $offset || DEFAULT_ITEMS_OFFSET;

        my $response = $self->_request($resource, $id_or_name, $queries);

        return WebService::Pokemon::APIResourceList->new(
            api => $self,
            response => $response
        );
    }

    my $response = $self->_request($resource, $id_or_name, $queries);

    return WebService::Pokemon::NamedAPIResource->new(
        api => $self,
        response => $response
    );
}

sub resource_by_url {
    my ($self, $url) = @_;

    my $uri = uri($url);

    my ($resource, $id_or_name);

    my $split_path = $uri->split_path;
    if (scalar @{$split_path} == 3) {
        $resource = @{$split_path}[-1];
    }
    else {
        $resource = @{$split_path}[-2];
        $id_or_name = @{$split_path}[-1];
    }

    return $self->resource(
        $resource, $id_or_name,
        $uri->param('limit'),
        $uri->param('offset'));
}

1;
__END__

=encoding utf-8

=for stopwords pokemon pokémon pokeapi autoload

=head1 NAME

WebService::Pokemon - Perl library for accessing the Pokémon data,
http://pokeapi.co.

=head1 SYNOPSIS

    use WebService::Pokemon;

    my $pokemon_api = WebService::Pokemon->new;

    # By id.
    my $pokemon = $pokemon_api->resource('berry', 1);

    # By name.
    my $pokemon = $pokemon_api->resource('berry', 'cheri');

=head1 DESCRIPTION

WebService::Pokemon is a Perl client helper library for the Pokemon API (pokeapi.co).

=head1 DEVELOPMENT

Source repository at L<https://github.com/kianmeng/webservice-pokemon|https://github.com/kianmeng/webservice-pokemon>.

How to contribute? Follow through the L<CONTRIBUTING.md|https://github.com/kianmeng/webservice-pokemon/blob/master/CONTRIBUTING.md> document to setup your development environment.

=head1 METHODS

=head2 new([%$args])

Construct a new WebService::Pokemon instance. Optionally takes a hash or hash reference.

    # Instantiate the class.
    my $pokemon_api = WebService::Pokemon->new;

=head3 api_url

The URL of the API resource.

    # Instantiate the class by setting the URL of the API endpoints.
    my $pokemon_api = WebService::Pokemon->new({api_url => 'http://example.com/api/v2'});

    # Or after the object was created.
    my $pokemon_api = WebService::Pokemon->new;
    $pokemon_api->api_url('http://example.com/api/v2');

=head3 autoload

Set this if you want to expand all fields point to external URL.

    # Instantiate the class by setting the URL of the API endpoints.
    my $pokemon_api = WebService::Pokemon->new({autoload => 1});

    # Or after the object was created.
    my $pokemon_api = WebService::Pokemon->new;
    $pokemon_api->autoload(1);
    $pokemon_api->resource('berry');

=head2 resource($resource, [$name], [$limit], [$offset])

Get the details of a particular resource with optional id or name; limit per
page, or offset by the record list.

    # Get paginated list of available berry resource.
    my $berry = $pokemon_api->resource('berry');

    # Or by page through limit and pagination.
    my $berry = $pokemon_api->resource('berry', undef, 60, 20);

    # Get by id.
    my $berry_firmness = $pokemon_api->resource('berry-firmnesses', 1);

    # Get by name.
    my $berry_firmness = $pokemon_api->resource('berry-firmnesses', 'very-soft');

=head2 resource_by_url($url)

Get the details of a particular resource by full URL.

    # Get paginated list of available berry resource with default item size.
    my $berries = $pokemon_api->resource_by_url('https://pokeapi.co/api/v2/berry/');

    # Get paginated list of available berry resource with explicit default item size.
    my $berries = $pokemon_api->resource_by_url('https://pokeapi.co/api/v2/berry/?limit=20&offset=40');

    # Get particular berry resource.
    my $berry = $pokemon_api->resource_by_url('https://pokeapi.co/api/v2/berry/1');

=head1 AUTHOR

Kian Meng, Ang E<lt>kianmeng@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2018-2019 by Kian Meng, Ang.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)

